src/audio/module_adapter/module_adapter.c

```diff
@@ -623,54 +630,126 @@ static void module_adapter_process_output(struct comp_dev *dev)
 	}
 }
 
+static uint32_t
+module_single_sink_setup(struct comp_dev *dev,
+			 struct comp_buffer __sparse_cache *source_c[PLATFORM_MAX_STREAMS],
+			 struct comp_buffer __sparse_cache *sinks_c[PLATFORM_MAX_STREAMS])
+{
+	struct processing_module *mod = comp_get_drvdata(dev);
+	struct comp_copy_limits c;
+	struct list_item *blist;
+	uint32_t num_input_buffers = 0;
+	int i = 0;
+
+	list_for_item(blist, &dev->bsource_list) {
+		/* check if the source dev is in the same state as the dev */
+		if (!source_c[i]->source || source_c[i]->source->state != dev->state) {
+			i++;
+			continue;
+		}
+
+		comp_get_copy_limits_frame_aligned(source_c[i], sinks_c[0], &c);
+
+		buffer_stream_invalidate(source_c[i], c.frames * c.source_frame_bytes);
+
+		/*
+		 * note that the size is in number of frames not the number of
+		 * bytes
+		 */
+		mod->input_buffers[num_input_buffers].size = c.frames;
+		mod->input_buffers[num_input_buffers].consumed = 0;
+
+		mod->input_buffers[num_input_buffers].data = &source_c[i]->stream;
```

Comment:
and here's why: to me this looks like your function receiver two arrays of pointers as parameters, i.e. it made local copies of those arrays. And here you store pointers to those (local copied of) array elements to global data. Maybe that isn't what the compiler does in this case, but it does look confusing to me.